File: R/simulate-data-1.R
# Load necessary libraries
library(tidyverse)
library(MASS) # For mvrnorm function

# Set seed for reproducibility
set.seed(1234)

# Parameters
n_students <- 60
n_assessments <- 6
n_items <- 20

# Generate group-level effects
group_effects <- tibble(
  Group = c("Control", "Self-Regulation"),
  group_initial_ability_mean = c(0.4, 0.42),
  group_improvement_rate_mean = c(0.05, 0.07)
)

# Generate student-level data with parameters drawn from the group-level effects
students <- tibble(
  StudentID = 1:n_students,
  Group = if_else(StudentID <= n_students / 2, "Control", "Self-Regulation")
) |>
  left_join(group_effects, by = "Group") |>
  rowwise() |>
  mutate(
    # Assume a correlation matrix and covariance matrix for our two random effects
    correlation_matrix = matrix(c(1, 0.6, 0.6, 1), 2), # Example correlation of 0.6 between effects
    sigma = matrix(c(0.05, 0, 0, 0.01), 2), # Variances for the random effects
    covariance_matrix = sigma %*% correlation_matrix %*% sigma,

    # Draw from a multivariate normal distribution
    effects = mvrnorm(1, c(group_initial_ability_mean, group_improvement_rate_mean), covariance_matrix),

    # Extract individual parameters
    individual_initial_ability = effects[1],
    individual_improvement_rate = effects[2]
  ) |>
  select(-group_initial_ability_mean, -group_improvement_rate_mean)

# Expand to create dataset with each assessment and item for each student
data <- students |>
  crossing(Assessment = 1:n_assessments, Item = 1:n_items) |>
  rowwise() |>
  mutate(
    prob = individual_initial_ability + Assessment * individual_improvement_rate,
    Score = rbinom(1, 1, prob)
  ) |>
  select(StudentID, Assessment, Item, Group, Score)

# View the first few rows of the data
head(data)

# Optional: Analysis can be done using multilevel modeling packages like `brms` or `lme4`.



# Compute average scores for each assessment and group
average_scores <- data |>
  group_by(Group, Assessment) |>
  summarize(avg_score = mean(Score), .groups = "drop")

# Plot the data using ggplot
ggplot(average_scores, aes(x = Assessment, y = avg_score, color = Group, group = Group)) +
  geom_line(aes(linetype = Group), size = 1) +
  geom_point(size = 3) +
  labs(
    title = "Average Score by Assessment and Group",
    x = "Assessment",
    y = "Average Score",
    color = "Group"
  ) +
  theme_minimal() +
  scale_color_manual(values = c("Control" = "blue", "Self-Regulation" = "red"))

File: jl/simulate-data-1.jl
using DataFrames, Random, Distributions, Plots, StatsBase

# Set seed for reproducibility
Random.seed!(1234)

# Parameters
n_students = 60
n_assessments = 6
n_items = 20
practice_amount = 5.0  # Example value: amount of practice every assessment
group_multiplier = 1.2  # Self-regulation group practices more effectively

# Create student-level data
students = DataFrame(
    StudentID = 1:n_students,
    Group = vcat(fill("Control", div(n_students,2)), fill("Self-Regulation", div(n_students,2))),
    PracticeEffectiveness = [i == "Self-Regulation" ? group_multiplier : 1.0 for i in vcat(fill("Control", div(n_students,2)), fill("Self-Regulation", div(n_students,2)))],
    InitialAbility = rand(Normal(10, 2), n_students)
)

# Define state transition function
function transition_ability(prev_ability, practice_effectiveness)
    return prev_ability + practice_amount * practice_effectiveness + rand(Normal(0, 1))
end

# Generate the state-space model data for each student and each assessment
abilities = []
scores = []
studentIDs = []
groups = []
assessments = []

for student in eachrow(students)
    student_ability = student.InitialAbility
    for a in 1:n_assessments
        push!(assessments, a)
        push!(studentIDs, student.StudentID)
        push!(groups, student.Group)
        
        student_ability = transition_ability(student_ability, student.PracticeEffectiveness)
        push!(abilities, student_ability)
        
        score_prob = cdf(Normal(15, 3), student_ability)
        score = sum(rand(Bernoulli(score_prob), n_items))
        push!(scores, score)
    end
end

data = DataFrame(StudentID=studentIDs, Assessment=assessments,




## without loop

using DataFrames, Random, Distributions, Plots, StatsBase

# Set seed for reproducibility
Random.seed!(1234)

# Parameters
n_students = 60
n_assessments = 6
n_items = 20
practice_amount = 5.0
group_multiplier = 1.2

# Create student-level data
students = DataFrame(
    StudentID = 1:n_students,
    Group = vcat(fill("Control", div(n_students,2)), fill("Self-Regulation", div(n_students,2))),
    PracticeEffectiveness = [i == "Self-Regulation" ? group_multiplier : 1.0 for i in vcat(fill("Control", div(n_students,2)), fill("Self-Regulation", div(n_students,2)))],
    InitialAbility = rand(Normal(10, 2), n_students)
)

# Define state transition function
function transition_ability(prev_ability, practice_effectiveness)
    return prev_ability + practice_amount * practice_effectiveness + rand(Normal(0, 1))
end

# Generate the state-space model data for each student and each assessment
studentIDs = []
groups = []
assessments = []
scores = []

for student in eachrow(students)
    abilities = reduce((acc, pe) -> vcat(acc, transition_ability(acc[end], pe)), 
                       fill(student.PracticeEffectiveness, n_assessments - 1), 
                       init=[student.InitialAbility])
    
    for (a, ability) in enumerate(abilities)
        push!(assessments, a)
        push!(studentIDs, student.StudentID)
        push!(groups, student.Group)
        
        score_prob = cdf(Normal(15, 3), ability)
        score = sum(rand(Bernoulli(score_prob), n_items))
        push!(scores, score)
    end
end

data = DataFrame(StudentID=studentIDs, Assessment=assessments, Group=groups, Score=scores)

# Plot the data
p = plot(groupby(data, [:Assessment, :Group]), 
    x=:Assessment, 
    y=:Score_mean, 
    line=(:Group, [:solid, :dash]), 
    color=:Group, 
    legend=:topright, 
    ylabel="Score", 
    xlabel="Assessment", 
    title="Score by Assessment and Group"
)
display(p)
